Added basic support for multiple outputs

DanChianucci opened the pull request:
Networks with multiple outputs can now be defined simply by passing a list of Layer instances as the layers parameter.  The networks _output_layer is set to the last layer in the list

Pertains to dnouri/nolearn#202

dnouri commented:
Thanks, this looks nice and simple.  Do you have a somewhat realistic usage example maybe?  Anything that we could point people to when they ask, how do I build multi-output networks.

Also a test would be nice.

DanChianucci commented:
Here is a quick gist showing how to use it.  Basically you just give a list of Layers instead of a single Layer.

With a custom objective, fit will correctly treat the network as having multiple outputs.
However, everything else (ie score, predict, custom_scores, ...) will treat the network as if it has a single output.  In this case the output layer is whatever the last Layer in the list was

https://gist.github.com/DanChianucci/c37520b6f3d391eae447221e03093a1b

As for tests, what types of things should I test for?  I'm thinking just initialize the network and make sure all of the layers are included in the layers_ attribute.  Anything else?

DanChianucci commented:
As for a realistic usage example, I could imagine trying to learn multiple attributes (ie. Animals w/ Attriutes, UCSD Birds, etc) with a single network.

A shared set of base layers could feed into individual attribute classifiers, and the objective would take care of training for all of them at once.

dnouri commented:
> As for tests, what types of things should I test for? I'm thinking just initialize the network and make sure all of the layers are included in the layers_ attribute. Anything else?

That sounds fine.

> However, everything else (ie score, predict, custom_scores, ...) will treat the network as if it has a single output

I'm noticing your gist does not include a way to predict and get outputs for task 1.  Is there a way you could extend your example to do that?

DanChianucci commented:
It looks like the failures were caused by MNIST not downloading properly.  The test I added seems to have passed.

I have also added an example of how to get outputs from multiple layers to my gist

dnouri commented on nolearn/lasagne/tests/test_base.py:
```diff
@@ -829,3 +832,44 @@ def test_it(self, grad_scale, layer):
+                        objective=objective)
+        net.initialize()
+
+        expected_names = sorted(["input", "conv1", "conv2", "hidden_a", "output_a", "hidden_b", "output_b"])
```
Looks like this is failing the pep8 test with Travis.

dnouri commented:
Thanks, this is looking great.  I added a little comment about a test failure.  Looks like Python 2.7 has another weirdo test failure which I'll look into tomorrow.

I looked at your updated gist and thought maybe there's a way to support multi-output predictions in the `NeuralNet.predict` method itself?  Not necessary to have for this pull request but would be interested to hear your thoughts.

BenjaminBossan commented:
Just my 2 cents:

I think this is an important addition.

sklearn has some support for "multioutput" estimators ([link](http://scikit-learn.org/stable/modules/multiclass.html)). E.g., you can pass a 2d array to RandomForestClassifier and it will switch to multioutput classification. I believe that when you call `predict_proba` on such a classifier, it will return a list of numpy arrays.

I know that the same scope is not exactly the same here, but it might be worth thinking about implementing similar interfaces. E.g., we could require one `y` for each output layer. (And possibly support adding a list of `objective`s for each output and each `y`.)

Being closer to sklearn would facilitate the usage of this feature, which I could imagine people would have difficulties using as of right now.

DanChianucci commented:
As to incorporating this into network.predict directly I can see a couple of possibilities.

I would recomend having `initialize_layers` return a list of the output layers.  This would in turn set `network._output_layer` to a list of layers.  

Then in _create_iter_funcs I would change `output_layer = layers[-1]` to `output_layer=self._output_layer`.  This in turn would make `predict_proba` get the output for all outputs.  `predict_proba` is only passed to the `scores_train`, `scores_valid` , and the `predict_iter_` functions.  

The user would need to take care to define their score functions accordingly.  Internally, `predict_iter_` is only used in `NeuralNet.predict_proba()`.  This could be refactored into something like 

    probas = []
    for Xb, yb in self.batch_iterator_test(X):
        probas.append(self.apply_batch_func(self.predict_iter_, Xb))
    
    # At this point probas has structure below
    #[(out1a,out2a,out3a,...),       #list of outputs for batch 1
    # (out1b,out2b,out3b,...), ...]  #list of outputs for batch 2

    probas = zip(*probas)
    #Here probas is a tuple of outputs
    #([out1a,out1b,...],        #list of batches from output 1
    # [out2a,out2b,...],        #list of batches from output 2
    # [out3a,out3b,...],...)    # ...
   
    output = (np.vstack(o) for o in probas)
    return output

This would behave the same as it currently does if a single output is specified, however if multiple outputs are passed to the constructor then the behavior would be for `predict_probas`, and in turn `predict`, to return a list of np.arrays, one for each output.

I kind of hinted at this in my gist, but this might break if net.regression is False

DanChianucci commented:
@BenjaminBossan
I think it would also be possible to have an objective which is a able to take a dictionary for its  loss_functions, and target arguments.  The dictionary would map from the layer name to the objective target.

I haven't put a lot of thought into it, but this might be as simple as making BatchIterator.__iter__ use _sldict to slice the Yb rather than doing it directly

dnouri commented:
Okay I'm going ahead and merging this now.  We can build whatever enhancements on top of this.

@DanChianucci I like the idea of allowing target values to be a dict.  It's in line with how inputs X works right now.

Would be happy to see another pull request for the additional features discussed.
